File: packages/ios/ContentfulOptimization/Sources/ContentfulOptimization/Core/ResolvedOptimizedEntry.swift
import Foundation

/// The result of resolving an optimized entry. `entry` is a `CTEntry` — `getField`, not `as?`
/// casts on a raw map — regardless of whether the baseline passed to `resolveOptimizedEntry` was
/// a raw `[String: Any]` or a `Contentful.Entry`; both overloads wrap their result the same way.
public struct ResolvedOptimizedEntry {
    public let entry: CTEntry
    public let selectedOptimization: [String: Any]?
    public let optimizationContextId: String?
}

File: packages/android/ContentfulOptimization/src/main/kotlin/com/contentful/optimization/core/ResolvedOptimizedEntry.kt
package com.contentful.optimization.core

import com.contentful.optimization.contentful.CTEntry

public data class ResolvedOptimizedEntry(
    val entry: CTEntry,
    val selectedOptimization: Map<String, Any>?,
    val optimizationContextId: String? = null,
    val isEmptyVariant: Boolean = false,
) {
    internal companion object {
        @Suppress("UNCHECKED_CAST")
        fun fromBridgeResult(
            result: Map<String, Any>,
            baselineEntry: CTEntry,
        ): ResolvedOptimizedEntry {
            val entry = (result["entry"] as? Map<String, Any>)
                ?.let { CTEntry.from(it, fallback = baselineEntry) }
                ?: baselineEntry
            return ResolvedOptimizedEntry(
                entry = entry,
                selectedOptimization = result["selectedOptimization"] as? Map<String, Any>,
                optimizationContextId = result["optimizationContextId"] as? String,
                isEmptyVariant = result["isEmptyVariant"] == true,
            )
        }
    }
}
